library/std/src/sys_common/time.rs

```diff
@@ -0,0 +1,120 @@
+mod tests;
+
+#[cfg_attr(not(windows), allow(dead_code))]
+pub const MONOTONIZE_U64: bool = cfg!(target_has_atomic = "64");
+
+#[cfg_attr(not(windows), allow(dead_code))]
+pub fn monotonize_u64(tick: u64) -> Option<u64> {
+    if MONOTONIZE_U64 {
+        use crate::sync::atomic::AtomicU64;
+        use crate::sync::atomic::Ordering::Relaxed;
+
+        static LAST_NOW: AtomicU64 = AtomicU64::new(0);
+
+        return Some(LAST_NOW.fetch_max(tick, Relaxed).max(tick));
+    }
+
+    None
+}
+
+#[cfg_attr(any(not(unix), target_os = "ios", target_os = "macos"), allow(dead_code))]
+pub fn can_monotonize_u64x2() -> bool {
+    cfg!(target_has_atomic = "128") || have_cmpxchg16b()
+}
+
+#[cfg_attr(any(not(unix), target_os = "ios", target_os = "macos"), allow(dead_code))]
+fn have_cmpxchg16b() -> bool {
+    cfg_if! {
+        if #[cfg(target_arch = "x86_64")] {
+            return std_detect::is_x86_feature_detected!("cmpxchg16b");
+        } else {
+            return false
+        }
+    }
+}
+
+#[cfg_attr(any(not(unix), target_os = "ios", target_os = "macos"), allow(dead_code))]
+#[allow(unused_variables)]
+pub fn monotonize_u64x2(sec: u64, nsec: u64) -> Option<(u64, u64)> {
+    #[cfg(target_arch = "x86_64")]
+    if have_cmpxchg16b() {
+        let os_now = (sec as u128) << 64 | nsec as u128;
+        let monotonized = unsafe { monotonize_cmpxchg16b(os_now) };
+        return Some(((monotonized >> 64) as u64, monotonized as u64));
+    }
+
+    cfg_if! {
+        if #[cfg(target_has_atomic = "128")] {
+            use core::sync::atomic::AtomicU128;
+            use crate::sync::atomic::Ordering::Relaxed;
+
+            static LAST_NOW: AtomicU128 = AtomicU128::new(0);
+
+            let os_now = (sec as u128) << 64 | nsec as u128;
+            let monotonized = LAST_NOW.fetch_max(os_now, Relaxed).max(os_now);
+            return Some(((monotonized >> 64) as u64, monotonized as u64));
+        } else {
+            return None;
+        }
+    }
+}
+
+#[cfg_attr(any(not(unix), target_os = "ios", target_os = "macos"), allow(dead_code))]
+#[cfg(target_arch = "x86_64")]
+#[target_feature(enable = "cmpxchg16b")]
+// FIXME(#56798) inlining this on baseline x86_64 would fail to link __sync_val_compare_and_swap_16
+#[cfg_attr(not(target_feature = "cmpxchg16b"), inline(never))]
+unsafe fn monotonize_cmpxchg16b(os_now: u128) -> u128 {
+    use crate::ptr::addr_of_mut;
+    use crate::sync::atomic::Ordering::Relaxed;
+    use core::arch::x86_64::cmpxchg16b;
+    use core::intrinsics::{atomic_load_relaxed, prefetch_write_data};
+    use core::mem::transmute;
+
+    #[repr(align(16))]
+    struct AlignedInstant(u128);
+
+    static mut LAST_NOW: AlignedInstant = AlignedInstant(0);
+
+    let last_ptr = addr_of_mut!(LAST_NOW) as *mut u128;
+
+    // obtain exclusive access to the cache line since we'll immediately follow up the loads
+    // (which would result in shared ownership) with a cmpxchg16b that requires it.
+    prefetch_write_data(last_ptr, 0);
```

Comment:
Have you actually measured this and seen it improve performance?

Reply:
Yes, although results heavily depend on core-placement. When placing all 4 threads of the contended case on adjacent cores and hyperthreads it improved both the best and worst case of multiple runs and eliminated the variance too.